Add overview, how it works, and glossary docs (#15)
divviup/public-docs: docs/how-it-works.md

---
sidebar_position: 2
---

# How It Works

Learn how Divvi Up works and how it protects privacy.

## Breakdown

Divvi Up ingests measurements and aggregates them, while protecting the privacy
of each individual measurement. This can be broadly broken down into several
steps, described in the following sections.

<!-- TODO(inahga): Consider making this a more detailed `mermaid` diagram -->

![](../static/how-it-works.png#gh-dark-mode-only)
![](../static/how-it-works-light.jpg#gh-light-mode-only)

See the [Distributed Aggregation Protocol (DAP)][DAP] for a highly detailed
specification of the protocol that underpins Divvi Up.

### Step 0: Subscriber sets up Divvi Up

Subscribers are users of Divvi Up, such as app developers, who wish to collect
quantitative measurements about their end-user applications.

The subscriber chooses a DAP client library, which is used to interact with any
DAP-compatible measurements aggregator, such as the aggregators provided by
Divvi Up and its partners. Client libraries for [Rust
(janus_client)][janus_client] and [TypeScript (divviup-ts)][divviup-ts] are
available.

For each measurement the subscriber wishes to aggregate, the subscriber creates

> a task. A task encodes metadata about how the measurement will be made, which

Reviewer: nit: I wouldn't say "encode", because that implies some specific wire representation of a message, and that's not the sense of a task we're discussing here. Maybe "A task contains" or "A task consists of".

aggregators will be used, and which aggregation function to use. The task's
parameters are shared with both aggregators.

### Step 1: Client generates measurements

The subscriber writes and deploys app functionality that takes measurements
corresponding to each task. They invoke a DAP client library and provide some of
the task parameters to it, such as the task identifier, measurement type, and
the aggregator servers being used.

For each measurement, the subsequent steps are performed.

### Step 2: Client shards the measurement

The app uses the DAP client library to randomly shard the measurement into two
parts, known as report shares. One report share is for the leader aggregator,
and the other for the helper aggregator.

> Each aggregator advertises a public key corresponding to its private key. The
> client library encrypts each share with the respective aggregator's public key.

Reviewer: I feel like "public key" already implies the existence of a corresponding private key.
```suggestion
The client library encrypts each share with a public key advertised by the respective aggregator.
```

### Step 3: Client sends shares to the aggregators

The DAP client library sends each report share via HTTPS to the leader and

> helper aggregators.

Reviewer: This isn't quite correct, because the client doesn't transmit a report share directly to the helper. I see there's a note about this below, but why not just say this plainly in this paragraph?

Author: I went with removing the admonitions and plainly stating reality. It reads well enough IMO, but we should have a more precise diagram to accompany this doc. I can work on this in a future PR.

Aggregators reject out-of-range or invalid measurements, as defined in the task

> parameters. This is done without knowledge of the original measurement.

Reviewer: I think it's worth discussing that rejecting measurements requires the aggregators to talk to each other to jointly evaluate validity proofs, because if they didn't do that, then DAP wouldn't require the aggregators to communicate at all. They could just sum secret shares and emit aggregate shares.

Also, discussing this here suggests that aggregators reject bad measurements during the "Client sends shares to the aggregators step", which isn't really true. DAP uploads conclude before aggregators evaluate proofs. I think discussion of proof validation ought to move to step 4. Or if you prefer, introduce a new step for per-report proof validation before discussing aggregation of report shares.

Each aggregator is operated by different organizations. Often, Divvi Up will act
as the leader, while another organization operates as a helper.

<!-- TODO(#11): link BYOH guide here -->

Crucially, these organizations are non-colluding. They don't directly share
report shares or otherwise conspire to defeat the protocol. DAP guarantees that

> as long as at least one of the aggregators is operated honestly, the privacy of

Reviewer: I think "operated honestly" is too complicated of an idea to define. How about "at least one of the aggregators does not defect from the protocol"?

Author: I went with "as long as  at least one of the aggregators operates according to the protocol".

each individual measurement will be protected.

If you don't wish to use any of the already-available third-party organizations
that Divvi Up partners with, you can self-host your own helper aggregator.

:::note

Both shares are actually sent to the leader aggregator. The leader relays the
helper's report share to the helper. Because the helper's report share is
encrypted to the helper, the leader cannot read it and the non-collusion
property is preserved.

This is done to simplify client library implementation, reduce the number of
client network round-trips, and reduce the operational requirements of the
helper.

:::

### Step 4: Aggregators perform aggregation

> Both aggregators work together to compute an aggregate across a batch of report
> shares, without revealing the report share to each other and without any
> knowledge of the originally submitted measurement.

Reviewer: The arity of the sentence should be consistent 
```suggestion
Both aggregators work together to compute an aggregate across a batch of report
shares, without revealing the report shares to each other and without any
knowledge of the originally submitted measurements.
```
